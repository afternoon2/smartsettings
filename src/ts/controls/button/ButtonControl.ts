import { Control } from '../control/Control';
import {
  ButtonControlProps, ControlListener,
} from '../control/Control.types';
import { InternalState } from '../../root/RootNode';

import Styles from '../../../sass/button.sass';

export class ButtonControl extends Control {
  public controlElement: HTMLButtonElement;

  protected static defaultType: string = 'button';

  protected static template = (state: InternalState): string => `
    <button
      class="${Styles.button}"
      type="${state.type || ButtonControl.defaultType}"
      title="${state.name}"
      ${state.disabled ? 'disabled="true"' : ''}
    >
      ${state.text || ''}
    </button>
  `;

  private controlListener?: ControlListener;
  private sectionListener?: ControlListener;
  private panelListener?: ControlListener;

  constructor(props: ButtonControlProps) {
    super(props, ButtonControl.template);
    this.controlElement = this.element.querySelector('button') as HTMLButtonElement;
    this.controlListener = props.listener;
    this.sectionListener = props.sectionListener;
    this.panelListener = props.panelListener;
    this.bindActionListeners();
  }

  private bindActionListeners() {
    this.controlElement.addEventListener('click', () => {
      if (this.controlListener) {
        this.controlListener({
          targetId: this.state.id,
          key: '',
          value: '',
          listenerType: 'control'
        });
      }
      if (this.sectionListener) {
        this.sectionListener({
          targetId: this.state.id,
          key: '',
          value: '',
          listenerType: 'section',
        });
      }
      if (this.panelListener) {
        this.panelListener({
          targetId: this.state.id,
          key: '',
          value: '',
          listenerType: 'panel'
        });
      }
    });
  }
}
