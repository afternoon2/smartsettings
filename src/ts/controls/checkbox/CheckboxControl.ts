import cuid from 'cuid';

import { Control } from '../control/Control';
import { InternalState, Listener, ListenerUpdate, CheckboxControlOptions } from '../../types';

import Styles from '../../../sass/checkbox.sass';

export type CheckboxControlProps = {
  id: string,
  options: CheckboxControlOptions,
  parentElement: HTMLElement,
  sectionListener?: Listener,
  panelListener?: Listener,
};

export class CheckboxControl extends Control {
  public controlElement: HTMLInputElement;
  public readonly displayType = 'checkbox';

  protected static template = (state: InternalState): string => {
    const id: string = cuid();
    return `<label
      class="${Styles.checkbox}" for="${id}">
        <input
          id="${id}"
          type="checkbox"
          title="${state.name}"
          ${state.checked ? 'checked' : ''}
          ${state.disabled ? 'disabled' : ''}
          ${state.readonly ? 'readonly="true"' : ''}
        />
      </label>
    `;
  }

  constructor(props: CheckboxControlProps) {
    super({
      ...props,
      template: CheckboxControl.template,
    });
    this.controlElement = this.element.querySelector('[type="checkbox"]') as HTMLInputElement;
    this.listeners.set('checked', this.onCheck);
    this.bindActionListeners();
  }

  get checked(): boolean {
    return this.state.checked as boolean;
  }

  check() {
    this.state.checked = true;
  }

  uncheck() {
    this.state.checked = false;
  }

  toggle() {
    this.state.checked = !this.state.checked;
  }

  private onCheck: Listener = (update: ListenerUpdate) => {
    if (update.value === true) {
      this.controlElement.setAttribute('checked', 'true');
    } else {
      this.controlElement.removeAttribute('checked');
    }
  }

  private bindActionListeners() {
    this.controlElement.addEventListener('change', (event: Event) => {
      event.preventDefault();
      this.toggle();
    });
  }
}
