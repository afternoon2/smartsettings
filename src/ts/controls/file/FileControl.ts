import cuid from 'cuid';

import { Control } from '../control/Control';
import { InternalState, Listener, ListenerUpdate, FileControlOptions } from '../../types';

import Styles from '../../../sass/file.sass';

export type FileControlProps = {
  id: string,
  options: FileControlOptions,
  parentElement: HTMLElement,
  sectionListener?: Listener,
  panelListener?: Listener,
};

export class FileControl extends Control {
  public controlElement: HTMLInputElement;

  protected static template = (state: InternalState): string => {
    const id: string = cuid();
    return `<label
        class="${Styles.file}"
        for="${id}"
      >
        Upload...
        <input
          id="${id}"
          type="file"
          title="${state.name}"
          ${state.disabled ? 'disabled' : ''}
          ${state.readonly ? 'readonly="true"' : ''}
        />
      </label>
    `
  };

  constructor(props: FileControlProps) {
    super({
      ...props,
      template: FileControl.template,
    });
    this.controlElement = this.element.querySelector('[type="file"]') as HTMLInputElement;
    this.listeners.set('accept', this.onAcceptChange);
    this.bindActionListeners();
  }

  get accept(): string {
    return this.state.accept as string;
  }

  set accept(accept: string) {
    this.state.accept = accept;
  }

  get files(): FileList | null {
    return this.state.files as FileList;
  }

  private onAcceptChange: Listener = (update: ListenerUpdate) => {
    this.controlElement.setAttribute('accept', update.value as string);
  }

  private bindActionListeners() {
    this.controlElement.addEventListener('change', (event: Event) => {
      this.state.files = ((event as Event).target as HTMLInputElement).files as FileList;
    });
  }
}
