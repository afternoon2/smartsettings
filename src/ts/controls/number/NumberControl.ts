import { Control } from '../control/Control';
import {
  NumberControlProps, ControlListener, ControlListenerUpdate,
} from '../control/Control.types';
import { InternalState } from '../../root/RootNode';

import Styles from '../../../sass/number.sass';

export class NumberControl extends Control {
  public controlElement: HTMLInputElement;

  protected static template = (state: InternalState): string => `<input
      type="number"
      class="${Styles.number}"
      ${state.value ? `value="${state.value}"` : ''}
      ${typeof state.min === 'number' ? `min="${state.min}"` : ''}
      ${typeof state.max === 'number' ? `max="${state.max}"` : ''}
      ${typeof state.step === 'number' ? `step="${state.step}"` : ''}
      ${state.readonly ? 'readonly="true"' : ''}
      ${state.disabled ? 'disabled' : ''}
      ${state.placeholder ? `placeholder="${state.placeholder}"` : ''}
    />
  `;

  constructor(props: NumberControlProps) {
    super(props, NumberControl.template);
    this.controlElement = this.element.querySelector('[type="number"]') as HTMLInputElement;
    this.listeners.set('value', this.onValue);
    this.bindActionListeners();
  }

  set value(value: number) {
    this.state.value = value;
  }

  get value(): number {
    return this.state.value as number;
  }

  private onValue: ControlListener = (update: ControlListenerUpdate) => {
    this.controlElement.value = update.value as string;
  }

  private bindActionListeners = () => {
    this.controlElement.addEventListener('change', (e: Event) => {
      this.value = parseFloat((e.target as HTMLInputElement).value);
    });
  }
}