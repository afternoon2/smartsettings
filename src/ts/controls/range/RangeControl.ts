import { Control } from '../control/Control';
import {
  RangeControlProps, ControlListener, ControlListenerUpdate,
} from '../control/Control.types';
import { InternalState } from '../../root/RootNode';

import Styles from '../../../sass/range.sass';
import ControlStyles from '../../../sass/control.sass';

export class RangeControl extends Control {
  public controlElement: HTMLInputElement;

  protected static template = (state: InternalState): string => `<input
      type="range"
      class="${Styles.range}"
      ${state.value && typeof state.value === 'number' ? `value="${state.value}"` : ''}
      ${typeof state.min === 'number' ? `min="${state.min}"` : ''}
      ${typeof state.max === 'number' ? `max="${state.max}"` : ''}
      ${typeof state.step === 'number' ? `step="${state.step}"` : ''}
      ${state.readonly ? 'readonly' : ''}
      ${state.disabled ? 'disabled' : ''}
    />
  `;

  constructor(props: RangeControlProps) {
    super(props, RangeControl.template);
    this.controlElement = this.element.querySelector('[type="range"]') as HTMLInputElement;
    this.updateMain();
    this.listeners.set('value', this.onValue);
    this.bindActionListeners();
  }

  set value(value: number) {
    this.state.value = value;
  }

  get value(): number {
    return this.state.value as number;
  }

  private onValue: ControlListener = (update: ControlListenerUpdate) => {
    this.controlElement.value = update.value as string;
    (this.element.querySelector(`.${Styles.rangeValue}`) as HTMLSpanElement).textContent = update.value as string;
  }

  private bindActionListeners = () => {
    this.controlElement.addEventListener('input', (e: Event) => {
      this.value = parseFloat((e.target as HTMLInputElement).value);
    });
  }

  private updateMain = () => {
    const main = this.element.querySelector(`.${ControlStyles.control__content}`) as HTMLElement;
    main.classList.add(`${ControlStyles['control__content--range']}`);
    main.insertAdjacentHTML('beforeend', `<span class="${Styles.rangeValue}">
      ${typeof this.state.value === 'number' ? this.state.value : ''}
    </span>`);
  }
}