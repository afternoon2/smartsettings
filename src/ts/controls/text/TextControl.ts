import { Control } from '../control/Control';
import { InternalState, TextControlOptions, Listener, ListenerUpdate } from '../../types';

import Styles from '../../../sass/text.sass';

export type TextControlProps = {
  id: string,
  options: TextControlOptions,
  parentElement: HTMLElement,
  sectionListener?: Listener,
  panelListener?: Listener,
};

export class TextControl extends Control {
  public controlElement: HTMLInputElement;
  public readonly displayType = 'text';

  protected static template = (state: InternalState): string => `
    <input
      type="text"
      class="${Styles.text}"
      ${state.disabled ? 'disabled' : ''}
      ${state.value ? `value="${state.value}"` : ''}
      ${state.autocomplete ? 'autocomplete="true"' : ''}
      ${state.minLength ? `min-length="${state.minLength}"` : ''}
      ${state.maxLength ? `min-length="${state.maxLength}"` : ''}
      ${state.placeholder ? `placeholder="${state.placeholder}"` : ''}
      ${state.pattern ? `pattern="${state.pattern}"` : ''}
      ${state.readOnly ? 'readonly="true"' : ''}
    />
  `;

  constructor(props: TextControlProps) {
    super({
      ...props,
      template: TextControl.template
    });
    this.controlElement = this.element.querySelector('[type="text"]') as HTMLInputElement;
    this.listeners.set('value', this.onValue);
    this.bindActionListeners();
  }

  set value(value: string) {
    this.state.value = value;
  }

  get value(): string {
    return this.state.value as string;
  }

  private onValue: Listener = (update: ListenerUpdate) => {
    this.controlElement.value = update.value as string;
  }

  private bindActionListeners = () => {
    this.controlElement.addEventListener('input', (e: Event) => {
      this.value = (e.target as HTMLInputElement).value;
    });
  }
}
