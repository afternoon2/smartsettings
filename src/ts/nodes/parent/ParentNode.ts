import cuid from 'cuid';

import { RootNode } from '../../root/RootNode';
import { ButtonControl, ButtonControlProps } from '../../controls/button/ButtonControl';
import { TextControl, TextControlProps } from '../../controls/text/TextControl';
import { TextAreaControl, TextAreaControlProps } from '../../controls/textarea/TextAreaControl';
import { CheckboxControl, CheckboxControlProps } from '../../controls/checkbox/CheckboxControl';
import { FileControl, FileControlProps } from '../../controls/file/FileControl';
import { NumberControl, NumberControlProps } from '../../controls/number/NumberControl';
import { RangeControl, RangeControlProps } from '../../controls/range/RangeControl';
import { DropDownControl, DropDownControlProps } from '../../controls/dropdown/DropDown';
import { ColorControl, ColorControlProps } from '../../controls/color/Color';
import { SectionNode } from '../section/Section';
import {
  Listener,
  ListenerUpdate,
  InternalState,
  ControlOptions,
  ParentOptions,
} from '../../types';

import Base from '../../../sass/base.sass';

export type AnyControl = ButtonControl
| TextControl
| TextAreaControl
| CheckboxControl
| FileControl
| NumberControl
| RangeControl
| DropDownControl
| ColorControl;

export type ParentNodeProps = {
  id: string,
  options: ParentOptions,
  template: (state: InternalState) => string,
  parentElement: HTMLElement,
}
export abstract class ParentNode extends RootNode {
  public parentElement: HTMLElement;
  public element: HTMLElement;

  public abstract bodyElement: HTMLElement;
  public abstract headerElement: HTMLElement;

  protected state: InternalState;
  protected listeners: Map<string, Listener> = new Map();
  protected stateHandler: ProxyHandler<InternalState> = {
    set: this.createStateSetter(),
  };
  protected registry: Map<string, AnyControl | SectionNode> = new Map();

  constructor(props: ParentNodeProps) {
    super();
    this.state = this.createState(props.id, props.options, this.stateHandler);
    this.parentElement = props.parentElement;
    this.element = this.createRootElement(props.template(this.state));

    this.parentElement.appendChild(this.element);
    this.listeners.set('invisible', this.onInvisible);
    this.listeners.set('disabled', this.onDisabled);
    this.listeners.set('collapsed', this.onCollapsed);
  }

  get collapsed(): boolean {
    return Boolean(this.state.collapsed);
  }

  open() {
    this.state.collapsed = false;
  }

  close() {
    this.state.collapsed = true;
  }

  toggle() {
    this.state.collapsed = !this.state.collapsed;
  }

  abstract control(
    control: string,
    options: ControlOptions,
  ): AnyControl | null;

  protected createControl = (
    control: string,
    options: ControlOptions, 
    sectionListener?: Listener,
    panelListener?: Listener,
  ): AnyControl | null => {
    const id: string = cuid();
    const parentElement = this.bodyElement;
    const props = {
      id,
      options,
      parentElement,
      sectionListener,
      panelListener,
    };
    let instance;
    switch (control) {
      case 'button':
        instance = new ButtonControl(props as ButtonControlProps);
        break;
      case 'checkbox':
        instance = new CheckboxControl(props as CheckboxControlProps);
        break;
      case 'dropdown':
        instance = new DropDownControl(props as DropDownControlProps);
        break;
      case 'file':
        instance = new FileControl(props as FileControlProps);
        break;
      case 'number':
        instance = new NumberControl(props as NumberControlProps);
        break;
      case 'range':
        instance = new RangeControl(props as RangeControlProps);
        break;
      case 'text':
        instance = new TextControl(props as TextControlProps);
        break;
      case 'textarea':
        instance = new TextAreaControl(props as TextAreaControlProps);
        break;
      case 'color':
        instance = new ColorControl(props as ColorControlProps);
        break;
      default:
        instance = null;
        break;
    }
    if (instance !== null) {
      this.registry.set(id, instance);
    }
    return instance;
  }

  protected onCollapsed: Listener = (update: ListenerUpdate) => {
    const { classList } = this.bodyElement;
    const toggle = this.headerElement.querySelector('a') as HTMLAnchorElement;
    if (update.value === true) {
      if (!classList.contains(Base.hidden)) {
        classList.add(Base.hidden);
      }
      toggle.innerHTML = '&#9662;';
    } else {
      if (classList.contains(Base.hidden)) {
        classList.remove(Base.hidden);
      }
      toggle.innerHTML = '&#9656;';
    }
  }

  protected createRootElement(template: string): HTMLElement {
    const rootElement = document.createElement('div');
    rootElement.insertAdjacentHTML('beforeend', template);
    return rootElement;
  }

  protected fillInElement(className: string) {
    this.element.classList.add(className);
    this.element.setAttribute('id', this.id);
    this.onDisabled({
      targetId: this.id,
      key: 'disabled',
      value: this.state.disabled as boolean,
      listenerType: 'builtin',
    });
  }
}
