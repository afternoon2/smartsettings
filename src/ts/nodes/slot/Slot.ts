import { ParentNode, AnyControl } from '../parent/ParentNode';
import {
  InternalState,
  Listener,
  ControlOptions,
  ConfigControlNode,
  SlotOptions,
  SlotConfig,
} from '../../types';
import { Control } from '../../controls/control/Control';

import Base from '../../../sass/base.sass';
import Styles from '../../../sass/slot.sass';
import { SectionNode } from '../section/Section';

export type SlotProps = {
  id: string,
  options: SlotOptions,
  parentElement: HTMLElement,
  panelListener?: Listener,
};

export class SlotNode extends ParentNode {
  public headerElement: HTMLElement;
  public bodyElement: HTMLElement;
  public contentElement: HTMLElement;
  public readonly displayType: string = 'slot';

  protected static template = (state: InternalState): string => `<header
      class="${Styles.slot__header}"
    >
      <p class="${Styles.slot__name}">
        ${state.name}
      </p>
      <a class="${Styles.slot__link}">
        ${state.collapsed ? '&#9662;' : '&#9656;'}
      </a>
    </header>
    <main class="${Styles.slot__body}">
      <div class="${Styles.slot__content} ${state.collapsed ? Base.hidden : ''}"></div>
    </main>
  `;

  constructor(props: SlotProps) {
    super({
      ...props,
      template: SlotNode.template,
    });
    if (props.options.listener) {
      this.listeners.set('slot', props.options.listener);
    }
    if (props.panelListener) {
      this.listeners.set('panel', props.panelListener);
    }
    this.fillInElement(Styles.slot);
    this.headerElement = this.element.querySelector(`.${Styles.slot__header}`) as HTMLElement;
    this.bodyElement = this.element.querySelector(`.${Styles.slot__body}`) as HTMLElement;
    this.contentElement = this.element.querySelector(`.${Styles.slot__content}`) as HTMLElement;
    this.bindEventListeners();
    if (!props.options.collapsed) {
      this.setBodyPosition();
    }
  }

  control(
    control: string,
    options: ControlOptions,
  ) {
    return this.createControl(
      control,
      options,
      this.listeners.get('slot'),
      this.listeners.get('panel'),
      this.contentElement,
    );
  }

  set config(config: SlotConfig) {
    Object
      .values(config)
      .forEach((opts: ConfigControlNode) => {
        // @ts-ignore
        this.control(opts.displayType, opts);
      });
  }

  get config(): SlotConfig {
    return Object.fromEntries(
      Array.from(this.registry.entries())
        .map((entry: [string, SlotNode | SectionNode | AnyControl]) => {
          const control: Control = entry[1] as Control;
          const configNode: ConfigControlNode = {
            displayType: control.displayType,
            ...control.properties,
          };
          return [
            entry[0],
            configNode,
          ];
        }),
    );
  }

  find(name: string): AnyControl | undefined {
    return Array.from(this.registry.values())
      .find(
        (node: AnyControl | SlotNode | SectionNode) => node.name === name
      ) as AnyControl | undefined;
  }

  findById(id: string): AnyControl | undefined {
    return this.registry.get(id) as AnyControl | undefined;
  }

  remove(name: string) {
    this.registry.forEach((value: SlotNode | SectionNode | AnyControl) => {
      if (value.name === name) {
        const toRemove = value.element;
        this.contentElement.removeChild(toRemove);
        this.registry.delete(value.id);
      };
    });
  }

  removeById(id: string) {
    const toRemove = this.registry.get(id);
    if (toRemove) {
      this.contentElement.removeChild(toRemove.element);
      this.registry.delete(id);
    }
  }

  removeAll() {
    while (this.contentElement.firstChild) {
      this.contentElement.removeChild(
        this.contentElement.firstChild,
      );
    }
    this.registry.clear();
  }

  setListener(listener: Listener) {
    this.listeners.set('slot', listener);
  }

  private onHeaderClick = (event: Event) => {
    event.preventDefault();
    if (this.state.collapsed === true) {
      this.open();
      this.setBodyPosition();
    } else {
      this.close();
    }
  }

  private bindEventListeners() {
    this.headerElement.addEventListener('click', this.onHeaderClick);
  }

  private setBodyPosition() {
    const { bodyElement, contentElement } = this;
    const bodyWidth: number = contentElement.clientWidth;
    const { left } = SlotNode.offset(bodyElement);
    const right: number = document.documentElement.clientWidth - (left + bodyWidth);

    if (right > bodyWidth && left > bodyWidth) {
      contentElement.style.top = '-20px';
      contentElement.style.left = '100%';
    } else {
      if (left > bodyWidth) {
        contentElement.style.top = '-20px';
        contentElement.style.left = '100%';
      } else {
        if (right > bodyWidth) {
          contentElement.style.top = '-20px';
          contentElement.style.left = '-100%';
        } else {
          contentElement.style.top = '0px';
          contentElement.style.left = '0px';
        }
      }
    }
  }

  private static offset(element: HTMLElement): { [key: string]: number } {
    const rect: DOMRect = element.getBoundingClientRect();
    const scrollLeft: number = window.pageXOffset || document.documentElement.scrollLeft;
    const scrollTop: number = window.pageYOffset || document.documentElement.scrollTop;
    return {
      top: rect.top + scrollTop,
      left: rect.left + scrollLeft,
    };
  }
}
