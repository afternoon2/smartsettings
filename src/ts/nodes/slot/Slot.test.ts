import { SlotNode } from './Slot';
import { ButtonControl } from '../../controls/button/ButtonControl';
import { TextControl } from '../../controls/text/TextControl';
import { TextAreaControl } from '../../controls/textarea/TextAreaControl';
import { NumberControl } from '../../controls/number/NumberControl';
import { RangeControl } from '../../controls/range/RangeControl';
import { CheckboxControl } from '../../controls/checkbox/CheckboxControl';
import { FileControl } from '../../controls/file/FileControl';
import { SlotConfig } from '../../types';
import { AnyControl } from '../parent/ParentNode';

let slot: SlotNode;
let listener: any;

describe('Slot node', () => {
  beforeEach(() => {
    listener = jest.fn();
    slot = new SlotNode({
      id: 'id',
      options: {
        name: 'name',
        collapsed: false,
        disabled: false,
        invisible: false,
        listener,
      },
      parentElement: document.body,
    });
  });

  test('If it matches the snapshot', () => {
    expect(slot.element).toMatchSnapshot();
  });

  test('Open and close', () => {
    slot.close();
    expect(slot.collapsed).toBe(true);
    expect(slot.element).toMatchSnapshot();
    slot.open();
    expect(slot.collapsed).toBe(false);
    expect(listener).toHaveBeenCalledTimes(2);
    expect(slot.element).toMatchSnapshot();
    slot.toggle();
    expect(slot.collapsed).toBe(true);
    expect(slot.element).toMatchSnapshot();
  });

  test('Open and close by click', () => {
    (slot.headerElement.querySelector('a') as HTMLAnchorElement).click();
    expect(slot.collapsed).toBe(true);
    expect(slot.element).toMatchSnapshot();
  });

  test('Control creation', () => {
    const btn = slot.control('button', { name: 'Button' });
    const text = slot.control('text', {
      name: 'Text',
      placeholder: 'Write here!',
      value: 'text input',
    });
    const textarea = slot.control('textarea', {
      name: 'Textarea',
      placeholder: 'textarea',
      value: 'Textarea',
    });
    const range = slot.control('range', {
      name: 'Range',
      min: 0, max: 10, value: 1, step: 1,
    });
    const number = slot.control('number', {
      name: 'Number',
      min: 0, max: 10, value: 1, step: 1,
    });
    const checkbox = slot.control('checkbox', {
      name: 'Checkbox',
      checked: true,
    });
    const file = slot.control('file', {
      name: 'Checkbox',
    });
    const falsyControl = slot.control('falsy', {
      name: 'Falsy',
      min: 1, max: 10, step: 1, value: 10,
    });
    expect(btn).toBeInstanceOf(ButtonControl);
    expect(text).toBeInstanceOf(TextControl);
    expect(textarea).toBeInstanceOf(TextAreaControl);
    expect(number).toBeInstanceOf(NumberControl);
    expect(range).toBeInstanceOf(RangeControl);
    expect(checkbox).toBeInstanceOf(CheckboxControl);
    expect(file).toBeInstanceOf(FileControl);
    expect(falsyControl).toBe(null);
  });

  test('Remove methods', () => {
    slot.control('button', { name: 'Button' });
    slot.control('button2', { name: 'Button' });
    slot.remove('Button');
    expect(slot.element.querySelector('#button')).toBe(null);
    expect(slot.element.querySelector('#button2')).toBe(null);

    const btn = slot.control('button', { name: 'Button' }) as ButtonControl;

    slot.removeById(btn.id);
    expect(slot.element.querySelector('button')).toBe(null);

    slot.control('button', { name: 'Button' }) as ButtonControl;

    slot.removeAll();
    expect(slot.contentElement.childElementCount).toBe(0);
  });

  test('Set listener', () => {
    const newListener = jest.fn();
    slot.setListener(newListener);
    slot.close();
    expect(newListener).toHaveBeenCalledTimes(1);
  });

  test('Set/get config', () => {
    const config: SlotConfig = {
      buttonControl: {
        type: 'button',
        displayType: 'button',
        name: 'btn',
      },
      textControl: {
        displayType: 'text',
        value: 'Incididunt ipsum laborum fugiat',
        name: 'input',
      },
    };
    slot.config = config;
    const btn = slot.element.querySelector('button') as HTMLButtonElement;
    const input = slot.element.querySelector('input') as HTMLInputElement;
    expect(btn).not.toBe(null);
    expect(input).not.toBe(null);
    expect(slot.config).toBeInstanceOf(Object);
  });

  test('Find and find by name', () => {
    const text = slot.control('text', {
      name: 'Text',
      placeholder: 'Write here!',
      value: 'text input',
    });
    const textarea = slot.control('textarea', {
      name: 'Textarea',
      placeholder: 'textarea',
      value: 'Textarea',
    }) as AnyControl;

    expect(slot.find('Text')).toBe(text);
    expect(slot.findById(textarea.id)).toBe(textarea);
    expect(slot.find('invalid name')).toBe(undefined);
    expect(slot.findById('invalid id')).toBe(undefined);
  });
});

