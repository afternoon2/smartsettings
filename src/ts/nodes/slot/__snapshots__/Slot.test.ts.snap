// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slot node If it matches the snapshot 1`] = `
<div
  class="undefined"
  id="id"
>
  <header
    class="slot__header"
  >
    
      
    <p
      class="slot__name"
    >
      
        name
      
    </p>
    
      
    <a
      class="slot__link"
    >
      
        ▸
      
    </a>
    
    
  </header>
  
    
  <main
    class="slot__body"
  >
    
      
    <div
      class="slot__content "
      style="top: 0px; left: 0px;"
    />
    
    
  </main>
  
  
</div>
`;

exports[`Slot node Open and close 1`] = `
<div
  class="undefined"
  id="id"
>
  <header
    class="slot__header"
  >
    
      
    <p
      class="slot__name"
    >
      
        name
      
    </p>
    
      
    <a
      class="slot__link"
    >
      ▾
    </a>
    
    
  </header>
  
    
  <main
    class="slot__body hidden"
  >
    
      
    <div
      class="slot__content "
      style="top: 0px; left: 0px;"
    />
    
    
  </main>
  
  
</div>
`;

exports[`Slot node Open and close 2`] = `
<div
  class="undefined"
  id="id"
>
  <header
    class="slot__header"
  >
    
      
    <p
      class="slot__name"
    >
      
        name
      
    </p>
    
      
    <a
      class="slot__link"
    >
      ▸
    </a>
    
    
  </header>
  
    
  <main
    class="slot__body"
  >
    
      
    <div
      class="slot__content "
      style="top: 0px; left: 0px;"
    />
    
    
  </main>
  
  
</div>
`;

exports[`Slot node Open and close 3`] = `
<div
  class="undefined"
  id="id"
>
  <header
    class="slot__header"
  >
    
      
    <p
      class="slot__name"
    >
      
        name
      
    </p>
    
      
    <a
      class="slot__link"
    >
      ▾
    </a>
    
    
  </header>
  
    
  <main
    class="slot__body hidden"
  >
    
      
    <div
      class="slot__content "
      style="top: 0px; left: 0px;"
    />
    
    
  </main>
  
  
</div>
`;

exports[`Slot node Open and close by click 1`] = `
<div
  class="undefined"
  id="id"
>
  <header
    class="slot__header"
  >
    
      
    <p
      class="slot__name"
    >
      
        name
      
    </p>
    
      
    <a
      class="slot__link"
    >
      ▾
    </a>
    
    
  </header>
  
    
  <main
    class="slot__body hidden"
  >
    
      
    <div
      class="slot__content "
      style="top: 0px; left: 0px;"
    />
    
    
  </main>
  
  
</div>
`;
